src/software/amazon/ionschema/internal/util/RangeIonTimestampPrecision.kt

```diff
@@ -0,0 +1,93 @@
+package software.amazon.ionschema.internal.util
+
+import software.amazon.ion.IonList
+import software.amazon.ion.IonSymbol
+import software.amazon.ion.IonTimestamp
+import software.amazon.ion.Timestamp
+import software.amazon.ionschema.InvalidSchemaException
+
+/**
+ * Implementation of Range<IonTimestamp> that validates that the precision
+ * of a given timestamp is within the expected range.
+ *
+ * This implementation simply translates the allowed timestamp precisions
+ * to a Range<Int>.
+ *
+ * @see IonTimestampPrecision
+ */
+internal class RangeIonTimestampPrecision (
+        ion: IonList
+) : Range<IonTimestamp> {
+
+    private val delegate: Range<Int>
+
+    init {
+        if (!ion.hasTypeAnnotation("range")) {
```

Comment:
Can be moved to the `init` block of a base class containing the `ion` property so it can be shared among all `Range<>` implementations.